Collect and apply field maps

tsalo opened the pull request:
Closes #63. Once this is working, I'll copy over the changes to `fmripost_template`.

## Changes proposed in this pull request

- Collect field maps from the fMRIPrep derivatives.
- Warp the preprocessed field map to BOLD reference space.
- Add a new version of the BOLD volumetric transform workflow that uses the preprocessed field map directly instead of hte coefficient and reference files.

tsalo committed:
Draft method to collect field map.

tsalo committed:
Merge branch 'main' into fmaps

tsalo committed:
Merge branch 'main' into fmaps

tsalo committed:
Merge branch 'main' into fmaps

tsalo committed:
Pass functional cache into inputnode.

tsalo committed:
Work on using fmap directly.

tsalo committed:
Try fixing.

tsalo committed:
Update pyproject.toml

effigies commented on src/fmripost_aroma/workflows/resample.py:
```diff
@@ -0,0 +1,189 @@
+    if not fieldmap_id:
+        return workflow
+
+    fmap_select = pe.Node(
+        KeySelect(fields=['fmap'], key=fieldmap_id),
+        name='fmap_select',
+        run_without_submitting=True,
+    )
```
You're not passing in multiple, so I don't think you need this.

```suggestion
```

effigies commented on src/fmripost_aroma/workflows/base.py:
```diff
@@ -427,30 +476,42 @@ def init_single_run_wf(bold_file):
         bold_MNI6_wf.inputs.inputnode.target_mask = mni6_mask
         bold_MNI6_wf.inputs.inputnode.target_ref_file = mni6_mask
 
+        # Warp desc-preproc_fieldmap to boldref space
+        # Warp the mask as well
+        fieldmap_to_boldref = pe.Node(
+            ApplyTransforms(),
+            name='fieldmap_to_boldref',
+        )
```
This node needs to go into the `bold_MNI6_wf` because you want to resample from fmap to the MNI space.

effigies commented on src/fmripost_aroma/workflows/resample.py:
```diff
@@ -0,0 +1,189 @@
+        name='fmap_select',
+        run_without_submitting=True,
+    )
+    distortion_params = pe.Node(
+        DistortionParameters(metadata=metadata),
+        name='distortion_params',
+        run_without_submitting=True,
+    )
+    workflow.connect([
+        (inputnode, fmap_select, [
+            ('fmap', 'fmap_ref'),
+            ('fmap_id', 'keys'),
+        ]),
+        (inputnode, distortion_params, [('bold_file', 'in_file')]),
+        # Inject fieldmap correction into resample node
+        (fmap_select, resample, [('fmap', 'fieldmap')]),
+        (distortion_params, resample, [
+            ('readout_time', 'ro_time'),
+            ('pe_direction', 'pe_dir'),
+        ]),
+    ])  # fmt:skip
```
Here is the section where you need to resample the fieldmap from fmap space to target space, using the inverse boldref2fmap transform, in addition to the boldref2anat and anat2MNI transforms. Then you pass in the fieldmap in MNI space to `resample`.

tsalo committed:
Update src/fmripost_aroma/workflows/resample.py

Co-authored-by: Chris Markiewicz <markiewicz@stanford.edu>

tsalo committed:
Start addressing review.

effigies committed:
rf: Use ResampleSeries to resample fieldmap

effigies committed:
sty: cleanup

effigies reviewed (commented):
Pushed what I was thinking for `resample.py`. Feel free to revert.

effigies committed:
Merge branch 'main' into fmaps

tsalo committed:
Update src/fmripost_aroma/workflows/base.py

Co-authored-by: Chris Markiewicz <markiewicz@stanford.edu>

tsalo commented on src/fmripost_aroma/workflows/base.py:
```diff
@@ -304,14 +306,31 @@ def init_single_run_wf(bold_file):
+            strict=False,
+            **{'from': 'boldref'},
+        )
+        fmapid = config.execution.layout.get_file(fmap_file).entities['to']
```
Great point!

tsalo commented:
@effigies thanks for intervening directly on this. I started addressing your review last weekend but got drawn into other things. I'm hoping to test out the changes here soon.

effigies commented:
@tsalo is this working for you? I need to implement it in fMRIPrep...

tsalo commented:
I haven't had a chance to try it yet. I might be able to give it a try by the end of the week though.

effigies commented:
Well, I'll implement it over in fMRIPrep and we'll see how I do. Thanks for laying the groundwork!

tsalo committed:
Merge branch 'main' into fmaps
